service/prometheus/service.go

```diff
@@ -2,71 +2,230 @@ package prometheus
 }
 
 func New(config Config) (*Service, error) {
+	if config.K8sClient == nil {
+		return nil, microerror.Maskf(invalidConfigError, "config.K8sClient must not be empty")
+	}
 	if config.Logger == nil {
 		return nil, microerror.Maskf(invalidConfigError, "config.Logger must not be empty")
 	}
 
 	if config.Address == "" {
 		return nil, microerror.Maskf(invalidConfigError, "config.Address must not be empty")
 	}
-
-	u, err := url.ParseRequestURI(config.Address)
-	if err != nil {
-		return nil, microerror.Maskf(invalidConfigError, "config.Address is invalid: %s", err)
+	if config.ConfigMapKey == "" {
+		return nil, microerror.Maskf(invalidConfigError, "config.ConfigMapKey must not be empty")
+	}
+	if config.ConfigMapName == "" {
+		return nil, microerror.Maskf(invalidConfigError, "config.ConfigMapName must not be empty")
+	}
+	if config.ConfigMapNamespace == "" {
+		return nil, microerror.Maskf(invalidConfigError, "config.ConfigMapNamespace must not be empty")
 	}
-	u.Path = path.Join(u.Path, prometheusReloadPath)
 
 	service := &Service{
-		logger: config.Logger,
+		k8sClient: config.K8sClient,
+		logger:    config.Logger,
 
-		address: u.String(),
+		address:            config.Address,
+		configMapKey:       config.ConfigMapKey,
+		configMapName:      config.ConfigMapName,
+		configMapNamespace: config.ConfigMapNamespace,
 	}
 
 	return service, nil
 }
 
 func (s *Service) Reload() error {
-	s.logger.Log("debug", fmt.Sprintf("reloading prometheus config: %s", s.address))
+	reloadRequired, err := s.isReloadRequired()
+	if err != nil {
+		return microerror.Maskf(reloadError, err.Error())
+	}
+
+	if reloadRequired {
+		if err := s.reload(); err != nil {
+			return microerror.Mask(err)
+		}
+	}
+
+	return nil
+}
+
+func (s *Service) isReloadRequired() (bool, error) {
+	s.logger.Log("debug", fmt.Sprintf("checking if reload is required"))
+
+	configurationReloadCheckCount.Inc()
+
+	currentConfigMapData, err := s.getCurrentConfigMapData()
+	if err != nil {
+		return false, microerror.Mask(err)
+	}
+
+	currentConfiguration, err := s.getCurrentConfig()
+	if err != nil {
+		return false, microerror.Mask(err)
+	}
+
+	fmt.Printf("currentConfigMapData:\n%s\n", currentConfigMapData)
+	fmt.Printf("currentConfiguration:\n%s\n", currentConfiguration)
+
+	if currentConfigMapData != currentConfiguration {
+		configurationReloadRequiredCount.Inc()
+
+		s.logger.Log("debug", "configmap and current configuration do not match, reload required")
+		return true, nil
+	}
+
+	s.logger.Log("debug", "configmap and current configuration match, reload not required")
+	return false, nil
+}
+
+// getCurrentConfigMapData returns the configuration that is in the configmap.
+func (s *Service) getCurrentConfigMapData() (string, error) {
+	s.logger.Log("debug", fmt.Sprintf("fetching configmap: %s/%s", s.configMapNamespace, s.configMapName))
 
-	res, err := http.Post(s.address, "", nil)
+	configMap, err := s.k8sClient.CoreV1().ConfigMaps(s.configMapNamespace).Get(
```

Comment:
I think CM can be passed to Reload from ProcessUpdateState, as the CM is the watched object (I believe). That way you don't have to pass K8sClient dependency here.

Reply:
I tried to keep the reloaded self-contained, so it's easier to remove later.

Reply:
I don't see it be more self-contained. You just remove Reload call with it's argument. But I can understand you prefer it that way.